source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,512 @@
+// always in sorted order so that two nodes with the same children but inserted
+// in a different order have the same hash and so that the overloaded operator==
+// will return true. If the node is already in the cache return the cached
+// version instead.
+//
+// 4. The created DAG can then be simplified by
+// ScalarAnalysis::SimplifyExpression, implemented in
+// scalar_analysis_simplification.cpp. See that file for further information on
+// the simplification process.
+//
+
+namespace spvtools {
+namespace opt {
+
+uint32_t SENode::NumberOfNodes = 0;
+
+SENode* ScalarEvolutionAnalysis::CreateNegation(SENode* operand) {
+  if (operand->GetType() == SENode::Constant) {
+    return CreateConstant(-operand->AsSEConstantNode()->FoldToSingleValue());
+  }
+  std::unique_ptr<SENode> negation_node{new SENegative(this)};
+  negation_node->AddChild(operand);
+  return GetCachedOrAdd(std::move(negation_node));
+}
+
+SENode* ScalarEvolutionAnalysis::CreateConstant(int64_t integer) {
+  return GetCachedOrAdd(
+      std::unique_ptr<SENode>(new SEConstantNode(this, integer)));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzeMultiplyOp(
+    const ir::Instruction* multiply) {
+  assert(multiply->opcode() == SpvOp::SpvOpIMul &&
+         "Multiply node did not come from a multiply instruction");
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  SENode* op1 =
+      AnalyzeInstruction(def_use->GetDef(multiply->GetSingleWordInOperand(0)));
+  SENode* op2 =
+      AnalyzeInstruction(def_use->GetDef(multiply->GetSingleWordInOperand(1)));
+
+  return CreateMultiplyNode(op1, op2);
+}
+
+SENode* ScalarEvolutionAnalysis::CreateMultiplyNode(SENode* operand_1,
+                                                    SENode* operand_2) {
+  if (operand_1->GetType() == SENode::Constant &&
+      operand_2->GetType() == SENode::Constant) {
+    return CreateConstant(operand_1->AsSEConstantNode()->FoldToSingleValue() *
+                          operand_2->AsSEConstantNode()->FoldToSingleValue());
+  }
+
+  std::unique_ptr<SENode> multiply_node{new SEMultiplyNode(this)};
+
+  multiply_node->AddChild(operand_1);
+  multiply_node->AddChild(operand_2);
+
+  return GetCachedOrAdd(std::move(multiply_node));
+}
+
+SENode* ScalarEvolutionAnalysis::CreateSubtraction(SENode* operand_1,
+                                                   SENode* operand_2) {
+  // Fold if both operands are constant.
+  if (operand_1->GetType() == SENode::Constant &&
+      operand_2->GetType() == SENode::Constant) {
+    return CreateConstant(operand_1->AsSEConstantNode()->FoldToSingleValue() -
+                          operand_2->AsSEConstantNode()->FoldToSingleValue());
+  }
+
+  return CreateAddNode(operand_1, CreateNegation(operand_2));
+}
+
+SENode* ScalarEvolutionAnalysis::CreateAddNode(SENode* operand_1,
+                                               SENode* operand_2,
+                                               bool simplify) {
```

Comment:
It defaults to true, the reason we have it is that in the simplification process we want to create add nodes and then add more children to them after. When initially creating the nodes we only ever have two children. At the end of the simplify, however, the graph is in a much flatter form with the add node as the root of the expression which can have more than two children. This is to make it easier to traverse all the terms in the expression as you just need to iterate over the immediate children of that add.

Reply:
I just removed the simplify flag, it was only actually used in one place. The idea was that the function does the equivalent of 
```
add child
add child
simplify
```

Whereas we needed
```
add child
add child
add more children
simplify
```
So we made it an optional flag to actually simplify the add node with just two children as otherwise we would end up adding more children to a constant node. However it was only actually used in one place and it doesn't make the code any bigger or less clear to just do the above operations manually so I just did that.